pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -325,47 +325,67 @@ func (c *NPLController) addPodIPToCache(key, podIP string) {
 	c.podToIP[key] = podIP
 }
 
+func (c *NPLController) deletePodIPFromCache(key string) {
+	c.podIPLock.Lock()
+	defer c.podIPLock.Unlock()
+	delete(c.podToIP, key)
+}
+
+func (c *NPLController) deleteAllPortRulesIfAny(podIP string) error {
+	data := c.portTable.GetDataForPodIP(podIP)
+	for _, d := range data {
+		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // handleRemovePod removes rules from port table and
 // rules programmed in the system based on implementation type (e.g. IPTABLES).
 // This also removes pod annotation from pods that are not selected by service annotation.
-func (c *NPLController) handleRemovePod(key string, obj interface{}) error {
-	klog.Infof("Got delete event for Pod: %s", key)
+func (c *NPLController) handleRemovePod(key string) error {
+	klog.V(2).Infof("Got delete event for Pod: %s", key)
 	podIP, found := c.getPodIPFromCache(key)
 	if !found {
 		klog.Infof("IP address not found for Pod: %s", key)
 		return nil
 	}
-	data := c.portTable.GetDataForPodIP(podIP)
-	for _, d := range data {
-		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
-		if err != nil {
-			return err
-		}
-	}
 
-	if obj != nil {
-		pod := obj.(*corev1.Pod)
-		if _, exists := pod.Annotations[NPLAnnotationKey]; exists {
-			newPod := pod.DeepCopy()
-			removePodAnnotation(newPod)
-			return c.updatePodAnnotation(newPod)
-		}
+	if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
+		return err
 	}
+
+	c.deletePodIPFromCache(key)
+
 	return nil
 }
 
 // handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
 func (c *NPLController) handleAddUpdatePod(key string, obj interface{}) error {
 	newPod := obj.(*corev1.Pod).DeepCopy()
-	klog.Infof("Got add/update event for pod: %s", key)
+	klog.V(2).Infof("Got add/update event for Pod: %s", key)
 
 	podIP := newPod.Status.PodIP
 	if podIP == "" {
-		klog.Infof("IP address not found for pod: %s/%s", newPod.Namespace, newPod.Name)
+		klog.Infof("IP address not set for Pod: %s", key)
 		return nil
 	}
 	c.addPodIPToCache(key, podIP)
 
+	if !c.isNPLEnabledForServiceOfPod(obj) {
+		if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
```

Comment:
Wonder do we have more efficient way to check if a Pod needs cleanup?

Reply:
I'm sure some improvements are possible. Maybe we can try to have a bi-directional map from Service to Pod. This PR doesn't try to improve that part. I think the heaviest computation happens if a Service's selector is changed, but I am not sure this is very common. Also in the end there a limited number of Pods on each Node so I doubt things can blow up too much. If on top of that only a few Services use NodePortLocal, it shouldn't be much overhead.

Reply:
Your reasoning makes senses to me. I do not like to iterate all Pods for any Pod update, but you are right isNPLEnabledForServiceOfPod() seems much complexer. Maybe we should consider differentiate different Pod changes, and optimize based on the types of changes. Of course I do not mean any thing before 0.13.